File: frontend/src/services/animalSaude.ts
import api, { type ApiResponse } from "./api";

export const TIPOS_CASO_SAUDE = [
  "TRATAMENTO",
  "PREVENTIVO",
  "CIRURGIA",
  "OUTRO",
] as const;

export type TipoCasoSaude = (typeof TIPOS_CASO_SAUDE)[number];

export const TIPO_CASO_SAUDE_LABELS: Record<TipoCasoSaude, string> = {
  TRATAMENTO: "Tratamento",
  PREVENTIVO: "Preventivo",
  CIRURGIA: "Cirurgia",
  OUTRO: "Outro",
};

export const STATUS_CASO_SAUDE = ["ATIVO", "CONCLUIDO", "CANCELADO"] as const;

export type StatusCasoSaude = (typeof STATUS_CASO_SAUDE)[number];

export const STATUS_CASO_SAUDE_LABELS: Record<StatusCasoSaude, string> = {
  ATIVO: "Ativo",
  CONCLUIDO: "Concluído",
  CANCELADO: "Cancelado",
};

export type AnimalSaudeRegistro = {
  id: number;
  animal_id: number;
  tipo_caso: TipoCasoSaude | string;
  data_inicio: string;
  data_fim?: string | null;
  status: StatusCasoSaude | string;
  observacoes?: string | null;
  created_by?: number | null;
  created_at: string;
  updated_at?: string | null;
};

export type SaveAnimalSaudePayload = {
  tipo_caso: string;
  data_inicio: string;
  data_fim?: string | null;
  status: string;
  observacoes?: string | null;
};

export function animalSaudeListQueryKey(animalId: number) {
  return ["animais", animalId, "saude"] as const;
}

export function animalSaudeDetailQueryKey(animalId: number, saudeId: number) {
  return ["animais", animalId, "saude", saudeId] as const;
}

export async function listByAnimal(animalId: number): Promise<AnimalSaudeRegistro[]> {
  const { data } = await api.get<ApiResponse<AnimalSaudeRegistro[]>>(
    `/api/v1/animais/${animalId}/saude`
  );
  return data.data ?? [];
}

export async function getById(
  animalId: number,
  saudeId: number
): Promise<AnimalSaudeRegistro> {
  const { data } = await api.get<ApiResponse<AnimalSaudeRegistro>>(
    `/api/v1/animais/${animalId}/saude/${saudeId}`
  );
  if (!data.data) throw new Error("Resposta inválida");
  return data.data;
}

export async function create(
  animalId: number,
  payload: SaveAnimalSaudePayload
): Promise<AnimalSaudeRegistro> {
  const { data } = await api.post<ApiResponse<AnimalSaudeRegistro>>(
    `/api/v1/animais/${animalId}/saude`,
    payload
  );
  if (!data.data) throw new Error("Resposta inválida");
  return data.data;
}

export async function update(
  animalId: number,
  saudeId: number,
  payload: SaveAnimalSaudePayload
): Promise<AnimalSaudeRegistro> {
  const { data } = await api.put<ApiResponse<AnimalSaudeRegistro>>(
    `/api/v1/animais/${animalId}/saude/${saudeId}`,
    payload
  );
  if (!data.data) throw new Error("Resposta inválida");
  return data.data;
}

export async function remove(animalId: number, saudeId: number): Promise<void> {
  await api.delete(`/api/v1/animais/${animalId}/saude/${saudeId}`);
}

export function dateOnlyFromApi(value?: string | null): string {
  if (!value) return "";
  return value.slice(0, 10);
}

File: backend/internal/models/animal_saude.go
package models

import "time"

const (
	AnimalSaudeTipoTratamento = "TRATAMENTO"
	AnimalSaudeTipoPreventivo = "PREVENTIVO"
	AnimalSaudeTipoCirurgia   = "CIRURGIA"
	AnimalSaudeTipoOutro      = "OUTRO"
)

const (
	AnimalSaudeStatusAtivo     = "ATIVO"
	AnimalSaudeStatusConcluido = "CONCLUIDO"
	AnimalSaudeStatusCancelado = "CANCELADO"
)

type AnimalSaude struct {
	ID          int64      `json:"id" db:"id"`
	AnimalID    int64      `json:"animal_id" db:"animal_id"`
	TipoCaso    string     `json:"tipo_caso" db:"tipo_caso"`
	DataInicio  time.Time  `json:"data_inicio" db:"data_inicio"`
	DataFim     *time.Time `json:"data_fim,omitempty" db:"data_fim"`
	Status      string     `json:"status" db:"status"`
	Observacoes *string    `json:"observacoes,omitempty" db:"observacoes"`
	VacinaID                    *int64 `json:"vacina_id,omitempty" db:"vacina_id"`
	HormonioLactacaoAplicacaoID *int64 `json:"hormonio_lactacao_aplicacao_id,omitempty" db:"hormonio_lactacao_aplicacao_id"`
	CreatedBy   *int64     `json:"created_by,omitempty" db:"created_by"`
	CreatedAt   time.Time  `json:"created_at" db:"created_at"`
	UpdatedAt   *time.Time `json:"updated_at,omitempty" db:"updated_at"`
}

func ValidAnimalSaudeTipos() []string {
	return []string{
		AnimalSaudeTipoTratamento,
		AnimalSaudeTipoPreventivo,
		AnimalSaudeTipoCirurgia,
		AnimalSaudeTipoOutro,
	}
}

func IsValidAnimalSaudeTipo(v string) bool {
	for _, t := range ValidAnimalSaudeTipos() {
		if t == v {
			return true
		}
	}
	return false
}

func ValidAnimalSaudeStatus() []string {
	return []string{
		AnimalSaudeStatusAtivo,
		AnimalSaudeStatusConcluido,
		AnimalSaudeStatusCancelado,
	}
}

func IsValidAnimalSaudeStatus(v string) bool {
	for _, s := range ValidAnimalSaudeStatus() {
		if s == v {
			return true
		}
	}
	return false
}
